ec/src/hashing/map_to_curve_hasher.rs

```diff
@@ -0,0 +1,76 @@
+use crate::{hashing::*, AffineCurve};
+use ark_ff::Field;
+use ark_std::{marker::PhantomData, vec::Vec};
+
+/// Trait for mapping a random field element to a random curve point.
+pub trait MapToCurve<T: AffineCurve> {
+    fn new_map_to_curve(domain: &[u8]) -> Result<Self, HashToCurveError>
+    where
+        Self: Sized;
+    /// Map random field point to a random curve point
+    fn map_to_curve(&self, point: T::BaseField) -> Result<T, HashToCurveError>;
+}
+
+/// Trait for hashing messages to field elements
+pub trait HashToField<F: Field>: Sized {
+    fn new_hash_to_field(domain: &[u8], count: usize) -> Result<Self, HashToCurveError>;
+
+    fn hash_to_field(&self, msg: &[u8]) -> Result<Vec<F>, HashToCurveError>;
+}
+
+/// Helper struct that can be used to construct elements on the elliptic curve
+/// from arbitrary messages, by first hashing the message onto a field element
+/// and then mapping it to the elliptic curve defined over that field.
+pub struct MapToCurveBasedHasher<T, H2F, M2C>
+where
+    T: AffineCurve,
+    H2F: HashToField<T::BaseField>,
+    M2C: MapToCurve<T>,
+{
+    field_hasher: H2F,
+    curve_mapper: M2C,
+    _params_t: PhantomData<T>,
+}
+
+impl<T, H2F, M2C> HashToCurve<T> for MapToCurveBasedHasher<T, H2F, M2C>
+where
+    T: AffineCurve,
+    H2F: HashToField<T::BaseField>,
+    M2C: MapToCurve<T>,
+{
+    fn new(domain: &[u8]) -> Result<Self, HashToCurveError> {
```

Comment:
Can the domain be dynamic, or is it generally statically known? I.e. do we sometimes need to set it at run-time? I can imagine the answer being yes sometimes, e.g. `for ii in 0..100, HashToCurve("domain {i}")`.

However if not, then maybe we can rely on constants instead of needing to construct a new `Vec` for the struct.

Reply:
We can't think of an application that the domain needs to be dynamic. It has been the structure that @ValarDragon had put in place. We can go ahead and replace it by a constant and hand it to the field hasher as const generic, if that is what you are suggesting.

Another somehow related change to @ValarDragon structure, is that MapToCurve does not use the domain to decide which element should be mapped to which point on the curve, so @mmagician removed domian completely from MapToCurve in this [pull request](https://github.com/w3f/arkworks-algebra/pull/23). Could you verify that this make sense and that you have no objection to this?

Reply:
I approved the changes in that PR; the design looks good to me.